docs/sync-overview.md

```diff
@@ -0,0 +1,57 @@
+  - For components that use local encryption, `SyncManager` passes the local encryption key to their `SyncEngine`
+  - Finally, calls `sync_multiple()` function from the `sync15` crate, sending it the `SyncEngine` instances.  `sync_multiple()` then calls the `sync()` function for each individual `SyncEngine`
+
+### Sync engines
+  - [`SyncEngine`](https://github.com/mozilla/application-services/blob/main/components/support/sync15-traits/src/engine.rs) is defined in the `support/sync15-traits` crate and defines the interface for syncing a component.
+  - A new `SyncEngine` instance is created for each sync
+  - `SyncEngine.apply_incoming()` does the main work.  It is responsible for processing incoming records from the server in order to update the local records and calculating which local records should be synced back.
+
+## The `apply_incoming` pattern
+
+`SyncEngine` instances are free to implement `apply_incoming()` any way they want, but the most components follow a general pattern.
+
+### Database Tables
+
+   - The local table stores records for the local application
+   - The mirror table stores the last known record from the server
+   - The staging temporary table stores the incoming records that we're currently processing
+   - Each table contains a `guid` as its primary key.  A record will share the same `guid` for the local/mirror/staging table.
+
+### `apply_incoming` stages
+  - **stage incoming**: write out each incoming server record to the staging table
+  - **fetch states**: take the rows from all 3 tables and combine them into a single struct containing `Option`s for the local/mirror/staging records.
+  - **iterate states**: loop through each state, decide how to do change the local records, then execute that plan.
+    - **reconcile/plan**: For each state we create an action plan for it.  The action plan is a low-level description of what to change (add this record, delete this one, modify this field, etc).  Here are some common situations:
+       - **A record only appears in the staging table**.  It's a new record from the server and should be added to the local DB
+       - **A record only appears in the local table**.  It's a new record on the local instance and should be synced back to the serve
+       - **Identical records appear in the local/mirror tables and a newer record is in the staging table**.  The record was updated remotely and the changes should be propagated to the local DB.
+       - **A record appears in the mirror table and newer records appear in both the local and staging tables**.  The record was updated both locally and remotely and we should perform a 3-way merge.
+    - **apply plan**: After we create the action plan, then we execute it.
+  - **fetch outgoing**:
+     - Calculate which records need to be sent back to the server
+     - Update the mirror table
+     - Return those records back to the `sync15` code so that it can sync them with the server.
```

Comment:
s/sync them with/upload them to/ - "sync them" kinda implies some additional logic, but there's none - we just POST exactly what we have been given.

Then a final bullet: * The sync15 code returns the timestamp reported by the server in the POST response and hands it back to the engine. The engine persists this timestamp in the `metadata` table - the next sync will then use this timestamp to fetch records that have since been changed by other devices